pinecone/client_test.go

```diff
@@ -686,11 +581,590 @@ func (ts *ClientTests) TestConfigureIndexHitPodLimit() {
+		},
+		{
+			name:         "Confirm ptr is returned when provided (int)",
+			ptr:          78,
+			defaultValue: 92,
+			expected:     78,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result := valueOrFallback(tt.ptr, tt.defaultValue)
+			assert.Equal(t, tt.expected, result, "Expected result to be '%s', but got '%s'", tt.expected, result)
+		})
+	}
+}
+
+func TestNewClientUnit(t *testing.T) {
+	testCases := []struct {
+		name            string
+		apiKey          string
+		headers         map[string]string
+		expectedHeaders map[string]string
+		expectedErr     bool
+	}{
+		{
+			name:   "Custom headers provided",
+			apiKey: "test-api-key",
+			headers: map[string]string{
+				"Test-Header": "custom-header-value",
+			},
+			expectedHeaders: map[string]string{
+				"Api-Key":     "test-api-key",
+				"Test-Header": "custom-header-value",
+			},
+			expectedErr: false,
+		},
+		{
+			name:            "No headers provided",
+			apiKey:          "test-api-key",
+			headers:         nil,
+			expectedHeaders: map[string]string{"Api-Key": "test-api-key"},
+			expectedErr:     false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockNewClientParams := NewClientParams{
+				ApiKey:  tc.apiKey,
+				Headers: tc.headers,
+			}
+
+			client, err := NewClient(mockNewClientParams)
+
+			if tc.expectedErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+				assert.NotNil(t, client)
+				assert.Equal(t, tc.expectedHeaders, client.headers, "Expected headers to be '%v', but got '%v'", tc.expectedHeaders, client.headers)
+			}
+		})
+	}
+}
+
+func TestNewClientBaseUnit(t *testing.T) {
+	// Save the current environment variable value and defer restoring it
+	originalHostEnv := os.Getenv("PINECONE_CONTROLLER_HOST")
+	defer os.Setenv("PINECONE_CONTROLLER_HOST", originalHostEnv)
+
+	testCases := []struct {
+		name         string
+		host         string
+		envHost      string
+		expectedHost string
+		expectedErr  bool
+	}{
+		{
+			name:         "Host provided in parameters",
+			host:         "https://custom-host.com/",
+			envHost:      "",
+			expectedHost: "https://custom-host.com/",
+			expectedErr:  false,
+		},
+		{
+			name:         "Host from environment variable",
+			host:         "",
+			envHost:      "https://env-host.com/",
+			expectedHost: "https://env-host.com/",
+			expectedErr:  false,
+		},
+		{
+			name:         "Default host",
+			host:         "",
+			envHost:      "",
+			expectedHost: "https://api.pinecone.io/",
+			expectedErr:  false,
+		},
+		{
+			name:         "Invalid URL scheme",
+			host:         "invalid-host			", // invalid b/c tab chars in url
+			envHost:      "",
+			expectedHost: "",
+			expectedErr:  true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Set environment variable for the test case
+			os.Setenv("PINECONE_CONTROLLER_HOST", tc.envHost)
```

Comment:
Should this only be set if `tc.envHost != ""`?

Reply:
no because then the 3rd test fails, since we set the env var `"PINECONE_CONTROLLER_HOST"` to whatever the value is of `tc.envHost`.

For the 3rd test, `tc.envHost` is equal to `""`. So then when we asserts that `tc.expectedHost` is equal to `client.restClient.Server`, it fails because `""` is not equal to `https://api.pinecone.io/`